### PR TITLE
Make `kl_div` a composite function.

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80334

Benchmarks: https://github.com/pytorch/pytorch/pull/80334#issuecomment-1167229285

Fixes https://github.com/pytorch/pytorch/issues/80158
Fixes https://github.com/pytorch/pytorch/issues/78867
Fixes https://github.com/pytorch/pytorch/issues/69230

Supersedes https://github.com/pytorch/pytorch/pull/79007
Supersedes https://github.com/pytorch/pytorch/pull/69212
Supersedes https://github.com/pytorch/pytorch/pull/19659

### CODE REVIEW

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -187,46 +187,12 @@ Tensor margin_ranking_loss(const Tensor& input1, const Tensor& input2, const Ten
...
-
-  if (reduction == at::Reduction::Mean) {
-    return grad_input / input.numel();
-  }
-  return grad_input;
+  TORCH_CHECK(!input.is_complex() && !target.is_complex(),
+              "kl_div: Complex inputs not supported.")
+  auto output = log_target ? at::exp(target) * (target - input)
+                           : at::xlogy(target, target) - target * input;
+  return apply_loss_reduction(output, reduction);
```
Comment (nikitaved) [inline aten/src/ATen/native/Loss.cpp:195]: We should broadcast inputs and add tests for these cases. See: https://github.com/pytorch/pytorch/pull/79007.

Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:195]: The inputs are broadcasted implicitly here. Now, it's not clear to me that we even want to support broadcasting. I think it would be better even to assert that `input.shape == target.shape`. Thoughts @jbschlosser?

Comment (nikitaved) [reply aten/src/ATen/native/Loss.cpp:195]: `kl_div` is not necessarily used in the loss function context. See https://github.com/pytorch/pytorch/issues/78867#issuecomment-1148981290

**Comment by Lezcano** [comment]
The benchmarks for a tensor of 1M elements. 

```
[--------- kl_div cpu --------]
               |   PR |  master
4 threads: --------------------
      fwd      |  1.1 |  0.7
      bwd      |  2.7 |  1.7
      fwd+bwd  |  3.6 |  2.6

Times are in milliseconds (ms).

[-------- kl_div cuda --------]
               |   PR |  master
4 threads: --------------------
      fwd      |  132 |  210
      bwd      |  523 |  297
      fwd+bwd  |  658 |  509

Times are in microseconds (us).
```

We see that we lose a bit of efficiency in the forward (which was expected). In return, we get correctness from many angles:
- This function is now correct with negative targets (returns -inf), with a target of zero (returns 0)
- Broadcasting is correct
- The implementation is now 2 lines long, supports meta tensors, etc

The bencharking script:
```python
from itertools import product

import torch
from torch.utils.benchmark import Compare, Timer


def get_timer(size, device, fn, name):
    # a is in the log-space, i.e. [-inf, 0]
    # b is a probability measure, i.e. [0, 1]
    a = torch.randn(*size, device=device)
    a = -torch.nn.functional.softplus(a)
    a.requires_grad_()

    b = torch.randn(*size, device=device).sigmoid()
    b.requires_grad_()

    z = torch.nn.functional.kl_div(a, b, reduction="batchmean")
    print(name, a.shape, device)

    timer = Timer(
        fn,
        globals={"a": a, "b": b, "z": z},
        label=f"kl_div {device}",
        sub_label=name,
        description="master",
        num_threads=4)

    return timer.blocked_autorange(min_run_time=5)


def get_params():
    fns = [("fwd", "torch.nn.functional.kl_div(a, b, reduction='batchmean')"),
           ("bwd", "z.backward(retain_graph=True)"),
           ("fwd+bwd", "torch.nn.functional.kl_div(a, b, reduction='batchmean').backward()")]
    for device, (name, fn) in product(("cpu", "cuda"), fns):
        yield (1000, 1000), device, fn, name


compare = Compare([get_timer(*params) for params in get_params()])
compare.trim_significant_figures()
compare.print()
```

**Comment by nikitaved** [comment]
I also expect the usage of `kl_div` in the context of multi-class classification with a small number of classes. Could you please benchmark these as well? With different batch sizes.

**Comment by Lezcano** [comment]
which sizes do you have in mind?

**Comment by Lezcano** [comment]
something like `(512, 50)`?

**Comment by nikitaved** [comment]
Since this composite solution is most likely still relying on the TI which has the following issue: https://github.com/pytorch/pytorch/issues/79218, could you please confirm there are no segfaults with the updated test cases? Provided that we keep the broadcasting behavior that is, CC @jbschlosser.

**Comment by nikitaved** [comment]
I think `n_classes = 2, 10, 50, and batch = 32, 64, 128, 512` should cover it, I hope.

File: aten/src/ATen/native/Loss.cpp [hunk 2]
```diff
@@ -187,46 +187,12 @@ Tensor margin_ranking_loss(const Tensor& input1, const Tensor& input2, const Ten
...
-  else {
-    grad_input = -at::exp(target) * grad_expand;
-  }
-
-  if (reduction == at::Reduction::Mean) {
-    return grad_input / input.numel();
-  }
-  return grad_input;
+  TORCH_CHECK(!input.is_complex() && !target.is_complex(),
+              "kl_div: Complex inputs not supported.")
```
Comment (nikitaved) [inline aten/src/ATen/native/Loss.cpp:192]: Makes perfect sense, we are working with probability measures here anyway.

**Comment by Lezcano** [comment]
Nevermind, I reverted the implementation to the previous one and now the new implementation is strictly better and faster than the previous one as it does strictly less operations. For example, for 1M elements:
```
[--------- kl_div cpu --------]
               |   PR |  master
4 threads: --------------------
      fwd      |  330 |  700
      bwd      |  280 |  1700
      fwd+bwd  |  720 |  2600

Times are in milliseconds (us).

[-------- kl_div cuda --------]
               |   PR |  master
4 threads: --------------------
      fwd      |  132 |  210
      bwd      |  263 |  297
      fwd+bwd  |  395 |  509

Times are in microseconds (us).
```

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (jbschlosser) [reply aten/src/ATen/native/Loss.cpp]: Even though I'd also like to see broadcasting disabled for losses, I don't think we can do that for `kl_div`. As @nikitaved mentioned, the function is also used for purposes beyond loss calculation.

**Comment by jbschlosser** [comment]
@Lezcano do we lose any numerical stability from making this composite? I'm seeing several NaNs show up in the failing checks

**Comment by Lezcano** [comment]
No we don't Those `NaNs` are from some tests that were not correct. The computations in this PR are pretty much the same as the ones that were there before.
The tests were corrected in https://github.com/pytorch/pytorch/pull/69212. I'll add the corrections to this PR.

**Comment by jbschlosser** [comment]
@Lezcano looks like we'll need to tweak `TestPythonDispatch.test_kwarg_only_and_positional_default` to use a different op now that `kl_div` gets decomposed. cc @ezyang hopefully this is okay

**Comment by ezyang** [comment]
@Lezcano and I discussed and we'll add a synthetic new op to test this

**Comment by Lezcano** [comment]
@jbschlosser @ezyang I added a `_foobar` operator for that test.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -13028,3 +13021,9 @@
   structured_inherits: TensorIteratorBase
   structured: True
   variants: function
+
+# Aux function used in the test TestPythonDispatch.test_kwarg_only_and_positional_default
+# within test/test_python_dispatch.py
+- func: _foobar(Tensor self, bool arg1=True, bool arg2=True, *, bool arg3=True) -> Tensor
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:13027]: nit: would be modestly better for the function name to say what its intended meaning is

**Review by ezyang — APPROVED** [review]

**Comment by nikitaved** [comment]
I guess we forgot to add tests for broadcasting, or did we not?

**Comment by Lezcano** [comment]
I did not add broadcasting tests in this PR. They can be added in a follow up PR

**Comment by nikitaved** [comment]
Right, but we need these tests to make sure that https://github.com/pytorch/pytorch/issues/78867 is indeed fixed.

**Comment by Lezcano** [comment]
I tested it locally and the provided example doesn't break in this branch anymore, so we're good here. We can add tests for this case later on to make sure we don't regress in the future.

**Comment by JackCaoG** [comment]
@Lezcano @ezyang Can we revert this pr, it actually break xla. XLA test was disabled last week due to some llvm download issue which might be why ci did not catch it. I will work on a fix...

**Comment by JackCaoG** [comment]
XLA still directly lowers https://github.com/pytorch/xla/blob/05fa2aa6ada872ab980e3fcbcac53644e4b09cf8/xla_native_functions.yaml#L164, I can try to remove it..

**Comment by JackCaoG** [comment]
if https://github.com/pytorch/xla/pull/3683 passed the build I will force merge that.. this way we don't need to revert and reland this pr.

**Comment by JackCaoG** [comment]
Thanks! I am working on a fix

**Comment by JackCaoG** [comment]
I kick of the build of https://github.com/pytorch/xla/pull/3683, if that passed I think we are good to reland.

**Comment by Lezcano** [comment]
Sorry I don't quite follow. Do we still need to manually update the XLA fix, or is it updated automatically after midnight, as it happened with the last PR we merged?

**Comment by JackCaoG** [comment]
@Lezcano Sorry for the confusion, so xla ci has been reenable in https://github.com/pytorch/pytorch/pull/81070#issuecomment-1178624356, so we have to follow the usual procedure. 

1. I will merge xla pr
2. You can rebase and update xla hash in .github/ci_commit_pins/xla.txt to new commit in xla
3. force merge this pr.

**Comment by Lezcano** [comment]
Fair enough. Merge that PR, ping me, and I'll update the hash and merge this one.

**Comment by Lezcano** [comment]
@JackCaoG just rebased this one

**Comment by CaoE** [comment]
kl_div will generate nan when target contain 0 values. For example:
```
in1 = torch.randn((n, n))
in2 = torch.tensor([[1, 0.5],[0.0, 0.6]])
out = torch.kl_div(in1, in2)
```
It may be a regression, not sure if it is due to this PR

**Comment by nikitaved** [comment]
@CaoE , it is a regression. The previous implementation applied `log` only on `target > 0`.

**Comment by Lezcano** [comment]
we can implement it via `xlogy` and we should be good. For that, we'd need to fix `xlogy`'s derivative to be more efficient, as it does way too many operations atm.

**Comment by Lezcano** [comment]
@nikitaved if you want, you can take on this one. I can review it in a week.

**Comment by nikitaved** [comment]
`xlogy` will not fix the issue with overflowing `exp`, unless there is a similar function like `xepxy`?

**Comment by Lezcano** [comment]
But it would resolve the regression pointed out in the comment above. Also, there's no issue with `exp` overflowing, as `exp` is just used when target is in a logarithmic scale, so all good thre as wel.

**Comment by PKizzle** [comment]
In the latest pytorch nightly version `torch-1.13.0.dev20220903` values of 0 for the target are not handled correctly on MPS. This might have something todo with MPS not currently supporting xlogy.

```
UserWarning: The operator 'aten::masked_select' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(
```

The resulting loss is nan.

```
tensor(nan, device='mps:0')
```

**Comment by Lezcano** [comment]
I could not fix the issue for MPS for that reason. See https://github.com/pytorch/pytorch/commit/c037103d1df61aaf078580d38dda8ced246764e6
Feel free to remove that path once MPS supports `xlogy`.
